### DISCUSSION TITLE
particle filtering in jupyter notebook

### DISCUSSION THREAD

**Discussion by amadan24** [zhonge/cryodrgn#392] [Q&A]
In particle filtering with the CryoDRGN Jupyter Notebook, after completing the low-resolution training and analysis, I am facing difficulty figuring out which particles/clusters to select - both with the lasso tool in the interactive section and in selecting a specific set of indices. Can anyone help me with this?

**Comment by michal-g** [comment]
Hey, just to clarify, are you having a technical issue with getting this tool to work properly? Or are you having difficulty identifying the correct subset of particles to select for further analyses?

In the case of the latter, we recommend looking at the volumes reconstructed at different locations within the latent space. If you have already run `cryodrgn analyze` with the default settings, there should be a `kmeans20` folder within your `analyze.x` output folder that contains such volumes (`vol_*.mrc`) with the locations chosen using k-means clustering of the locations of the particles within the latent space. Plots in the same folder such as `umap_hex.png` indicate where within the latent space each of these reconstruction lie, and matching better or worse reconstructions to areas or clusters within the latent space should give you some idea as to which indices to select.

It is also possible that in the case of your dataset you are not training a sufficiently expressive model to capture the signal present, or that you simply need to tune some hyper-parameters, in which case you should consider increasing the number of nodes in the model's hidden layers (`--enc-dim` and `--dec-dim`), as well as trying different values for learning rates (`--lr` and `--pose-lr`), and optimizer weight decay (`--wd`).

Let us know if this helps!

**Comment by amadan24** [comment]
Hi Michal,Thanks for your response. I’m not sure what you mean by matching better or worse reconstructions to clusters. Do you mind hopping on a zoom sometime this week to sort it out?Thanks,AnikaOn Aug 14, 2024, at 11:36 AM, Michal Grzadkowski ***@***.***> wrote:﻿
Hey, just to clarify, are you having a technical issue with getting this tool to work properly? Or are you having difficulty identifying the correct subset of particles to select for further analyses?
In the case of the latter, we recommend looking at the volumes reconstructed at different locations within the latent space. If you have already run cryodrgn analyze with the default settings, there should be a kmeans20 folder within your analyze.x output folder that contains such volumes (vol_*.mrc) with the locations chosen using k-means clustering of the locations of the particles within the latent space. Plots in the same folder such as umap_hex.png indicate where within the latent space each of these reconstruction lie, and matching better or worse reconstructions to areas or clusters within the latent space should give you some idea as to which indices to select.
It is also possible that in the case of your dataset you are not training a sufficiently expressive model to capture the signal present, or that you simply need to tune some hyper-parameters, in which case you should consider increasing the number of nodes in the model's hidden layers (--enc-dim and --dec-dim), as well as trying different values for learning rates (--lr and --pose-lr), and optimizer weight decay (--wd).
Let us know if this helps!

**Comment by michal-g** [comment]
I'm not going to be free to Zoom until next week, in the meantime can you share your `umap_hex.png` here and maybe I can explain it more effectively using this plot as a reference? Please include renderings of some of the `vol_*.mrc` files found in the same `analyze.x/kmeans20` output folder too if possible!

**Comment by amadan24** [comment]
Hi Michal,I will be in the lab tomorrow morning and can send all this information to you then. Are you able to zoom tomorrow? Thanks,AnikaOn Aug 20, 2024, at 7:21 PM, Michal Grzadkowski ***@***.***> wrote:﻿
I'm not going to be free to Zoom until next week, in the meantime can you share your umap_hex.png here and maybe I can explain it more effectively using this plot as a reference? Please include renderings of some of the vol_*.mrc files found in the same analyze.x/kmeans20 output folder too if possible!

**Comment by michal-g** [comment]
I won't be free for a Zoom until Wednesday (feel free to reach out via email to schedule), but please share your results here if possible!

**Comment by amadan24** [comment]
Ok, no worries. Here is the umap and some volumes we got:

 vol_002.mrc
<https://drive.google.com/file/d/1UFthLuJUwG3HJkJ4LbFJttt1R8f9M5bb/view?usp=drive_web>
 vol_006.mrc
<https://drive.google.com/file/d/1D1vgKQm1iLnNYe8CjABzTAnCr75gg255/view?usp=drive_web>
 vol_013.mrc
<https://drive.google.com/file/d/1wgYIPudaYR4fDXTn94A0qeukSPc32Yk6/view?usp=drive_web>
 vol_018.mrc
<https://drive.google.com/file/d/17c8R1cCsy1J4MXtg1lcS5t_jcZfEAc5B/view?usp=drive_web>
 vol_001.mrc
<https://drive.google.com/file/d/1WefpeLRBYCAbt_VrunE6XzZszH3E9EVT/view?usp=drive_web>

**Comment by michal-g** [comment]
Just requested access to these files! Did you mean to share a UMAP plot along with these volumes?

**Comment by amadan24** [comment]
Yes, here it is again:

**Comment by michal-g** [comment]
There is no plot being displayed when I view the message above, nor is there a link to a plot file — can you please double-check things on your end?

**Comment by amadan24** [comment]
Sorry about that! Does this work?

**Comment by michal-g** [comment]
It does not, can you just email me the file at mg2332@princeton.edu?

**Comment by michal-g** [comment]
*`umap_hexbin.png`*
![umap_hex (1)](https://github.com/user-attachments/assets/dd4409fb-0463-43d3-b2b7-ce0425539971)
*`vol_001.mrc`*
![image2](https://github.com/user-attachments/assets/d61f8705-98a6-4ab8-bbb6-64b3f61bb121)
*`vol_002.mrc`*
![image1](https://github.com/user-attachments/assets/743449fc-b9a2-4432-8f70-1b043fb0b2dd)

Across your volumes `vol_*.mrc`, are there any that look like junk particles? Are there any that instead look closer to the shape of the molecule you are hoping to reconstruct?

For example, between the two volumes that you shared with me, `vol_001.mrc` looks like mostly noise, while `vol_002.mrc` looks like it is starting to develop a distinct shape (ChimeraX visualizations shown above). Assuming your volume files are 0-indexed (go from `vol_000.mrc` to `vol_019.mrc`), that means point `1` labelled on the UMAP plot above corresponds to a region of the latent embedding space where there are likely many particles corresponding to such noise, while point `2` corresponds to a region of the latent embedding space where there are likely particles containing more signal (points `0` and `1` respectively if your volumes are 1-indexed and go from `vol_001.mrc` to `vol_020.mrc`).

After examining the other volumes to get a fuller understanding of which parts of the latent space correspond to good particles, you can use tools such as `cryodrgn filter` to select these regions to create indices that will subset the particles stack for future analyses. Assuming your volume files were 1-indexed, this would suggest selecting a region in the center of your latent embedding mass, thus excluding the areas around the edges containing particles such as `0` — after, of course, checking that points like `14`, `15`, and `16` are also bad particles while points like `8`, `9`, and `12` look better!
